recipes-app/iot2050-firmware-update/files/iot2050-firmware-update

```diff
@@ -218,6 +220,54 @@ class FirmwareUpdate(object):
 
             mtd_num += 1
 
+        self.firmware_integrity_check()
+
+    def flash_md5_digest(self, firmware_len, mtd_num=0):
+        flash_md5 = hashlib.md5()
+
+        while True:
+            if firmware_len <= 0:
+                break
+
+            mtd_dev_path, mtd_size, mtd_erasesize, mtd_name = self.get_mtd_info(
+                mtd_num)
+
+            mtd_pos = 0
+            try:
+                mtd_dev = os.open(mtd_dev_path, os.O_SYNC | os.O_RDWR)
+            except IOError as e:
+                print("Opening {} failed: {}".format(mtd_dev_path, e.strerror))
+                sys.exit(1)
+
+            while mtd_pos < mtd_size and firmware_len > 0:
+                mtd_content = os.read(mtd_dev, mtd_erasesize)
+                flash_md5.update(mtd_content)
+
+                mtd_pos += mtd_erasesize
+                firmware_len -= mtd_erasesize
+
+            os.close(mtd_dev)
+
+            mtd_num += 1
+
+        return flash_md5.hexdigest()
+
+    def firmware_md5_digest(self):
+        self.firmware.seek(0)
+        firmware_md5 = hashlib.md5()
+
+        for chunk in iter(lambda: self.firmware.read(4096), b""):
+            firmware_md5.update(chunk)
+
+        return firmware_md5.hexdigest()
+
+    def firmware_integrity_check(self):
+        firmware_md5_digest = self.firmware_md5_digest()
+        firmware_flash_md5_digest = self.flash_md5_digest(self.firmware_len)
+
+        if firmware_md5_digest != firmware_flash_md5_digest:
+            raise Exception('Upgrade Failed! Please do not reboot. Please try updating firmware again!\n')
+
```

Comment:
Per the convention used in other place of this script, I would rather use a print and a exit(!0).

A pythonic exception printing seems more like a internal error of the application, at least to me.

Or do we have somewhere in the script to capture this exception?

Reply:
By the way, better to inform the user if they reboot the device, it could be bricked.

Reply:
Just printing and then exiting, I'm afraid the user will ignore this message. At least we need obvious error messages

Reply:
For the prompt language, it can be optimized

Reply:
> Just printing and then exiting, I'm afraid the user will ignore this message. At least we need obvious error messages

We can use all CAPTION words, or triple `!!!` or a big `WARNING:`.

Reply:
Please also pay attention to something like `flash_erase`, such kind of error also makes the device un-bootable.

Reply:
After a closer look into the current code, the whole exception logic should be refactored to capture the programming failed event.

Better to throw all the exceptions to the most up level called and let the most upper level to handle them to report the warning.

Reply:
there is something wrong and this error is Fatal Error.
I think the better way is to abort and prompt. Just print is not enough